nodeup/pkg/model/docker.go

```diff
@@ -268,5 +270,103 @@ func (b *DockerBuilder) Build(c *fi.ModelBuilderContext) error {
 		}
 	}
 
+	dockerSemver, err := semver.Parse(dockerVersion)
+	if err != nil {
+		return fmt.Errorf("error parsing docker version %q as semver: %v", dockerVersion, err)
+	}
+
+	c.AddTask(b.buildSystemdService(dockerSemver))
+
 	return nil
 }
+
+func (b *DockerBuilder) buildSystemdService(dockerVersion semver.Version) *nodetasks.Service {
+	oldDocker := dockerVersion.Major <= 1 && dockerVersion.Minor <= 11
+	usesDockerSocket := true
+	hasDockerBabysitter := false
+
+	var dockerdCommand string
+	if oldDocker {
+		dockerdCommand = "/usr/bin/docker dameon"
+	} else {
+		dockerdCommand = "/usr/bin/dockerd"
+	}
+
+	if b.Distribution.IsDebianFamily() {
+		hasDockerBabysitter = true
+	}
+
+	manifest := &ServiceManifest{}
+	manifest.Set("Unit", "Description", "Docker Application Container Engine")
+	manifest.Set("Unit", "Documentation", "https://docs.docker.com")
+
+	if usesDockerSocket {
+		manifest.Set("Unit", "After", "network.target docker.socket")
+		manifest.Set("Unit", "Requires", "docker.socket")
+	} else {
+		manifest.Set("Unit", "After", "network.target")
+	}
+
+	manifest.Set("Service", "Type", "notify")
+	manifest.Set("Service", "EnvironmentFile", "/etc/sysconfig/docker")
+
+	if usesDockerSocket {
+		manifest.Set("Service", "ExecStart", dockerdCommand+" -H fd:// \"$DOCKER_OPTS\"")
+	} else {
+		manifest.Set("Service", "ExecStart", dockerdCommand+" \"$DOCKER_OPTS\"")
+	}
+
+	if !oldDocker {
+		// This was added by docker 1.12
+		// TODO: They seem sensible - should we backport them?
+
+		manifest.Set("Service", "ExecReload", "/bin/kill -s HUP $MAINPID")
+		// kill only the docker process, not all processes in the cgroup
+		manifest.Set("Service", "KillMode", "process")
+
+		manifest.Set("Service", "TimeoutStartSec", "0")
+	}
+
+	if oldDocker {
+		// Only in older versions of docker (< 1.12)
+		manifest.Set("Service", "MountFlags", "slave")
+	}
+
+	// Having non-zero Limit*s causes performance problems due to accounting overhead
+	// in the kernel. We recommend using cgroups to do container-local accounting.
+	// TODO: Should we set this?
+	//service.Set("Service", "LimitNOFILE", "infinity")
```

Comment:
I opened https://github.com/kubernetes/kubernetes/issues/39682 to figure out if we should set these to infinity